CSharp/Library/QnAMaker/QnAMakerDialog.cs

```diff
@@ -1,110 +1,225 @@
+            {
+                var tasks = this.services.Select(s => s.QueryServiceAsync(message.Text)).ToArray();
+
+                if (tasks.Any())
+                {
+                    var maxValue = tasks.Max(x => x.Result.Answers[0].Score);
+                    qnaMakerResults = tasks.First(x => x.Result.Answers[0].Score == maxValue).Result;
+
+                    var filteredQnAResults = new List<QnAMakerResult>();
+                    foreach (var qnaMakerResult in qnaMakerResults.Answers)
+                    {
+                        qnaMakerResult.Score /= 100;
+                        if (qnaMakerResult.Score >= qnaMakerResults.ServiceCfg.ScoreThreshold)
+                        {
+                            qnaMakerResult.Answer = HttpUtility.HtmlDecode(qnaMakerResult.Answer);
+                            filteredQnAResults.Add((qnaMakerResult));
+                        }
+                    }
+
+                    qnaMakerResults.Answers = filteredQnAResults;
+
+                    if (qnaMakerResults != null && qnaMakerResults.Answers != null && qnaMakerResults.Answers.Count > 0)
+                    {
+                        if (this.IsConfidentAnswer(qnaMakerResults))
+                        {
+                            await this.RespondFromQnAMakerResultAsync(context, message, qnaMakerResults);
+                            await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                        }
+                        else
+                        {
+                            var feedbackRecord = new FeedbackRecord { UserId = message.From.Id, UserQuestion = message.Text };
+                            context.PrivateConversationData.SetValue("qnaMakerResults", qnaMakerResults);
+                            context.PrivateConversationData.SetValue("feedbackRecord", feedbackRecord);
+                            await this.QnAFeedbackStepAsync(context, qnaMakerResults);
+                        }
+                    }
+                    else
+                    {
+                        await context.PostAsync(qnaMakerResults.ServiceCfg.DefaultMessage);
+                        await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                    }
+                }
+                else
+                {
+                    await context.PostAsync(qnaMakerResults.ServiceCfg.DefaultMessage);
+                    await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                }
+            }
+            else
+            {
+                await context.PostAsync(qnaMakerResults.ServiceCfg.DefaultMessage);
+                await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+            }
+        }
+
+        protected virtual bool IsConfidentAnswer(QnAMakerResults qnaMakerResults)
+        {
+            if (qnaMakerResults.Answers.Count < 2 || qnaMakerResults.Answers.FirstOrDefault().Score >= 99)

```

Comment:
I've added them as constants. 

They need not be configurable. This function helps is identifying the queries where QnA maker scores are good and the top answer can be directly returned. Since its a virtual function, one can override the function to add their own logic and thresholds.